src/uct/api/uct.h

```diff
@@ -665,13 +668,20 @@ void uct_iface_mem_free(const uct_allocated_memory_t *mem);
 
 /* @ingroup UCT_AM
  * @brief  List of capabilities of active message callback
+ *
+ * A callback must have either SYNC or ASYNC capability.
  */
 enum {
-    UCT_AM_CB_FLAG_DEFAULT     = 0,          /**< callback is always invoked from the thread 
-                                                  that called uct_iface_progress() */
-    UCT_AM_CB_FLAG_THREAD_SAFE = UCS_BIT(1), /**< callback may be invoked from any
-                                                  thread. For example it may be called
-                                                  from transport async progress thread */ 
+    UCT_AM_CB_FLAG_SYNC  = UCS_BIT(1), /**< callback is always invoked from the thread 
+                                            that called uct_iface_progress(). An interface must
+                                            have UCT_IFACE_FLAG_AM_CB_SYNC capability to support sync 
+                                            callback invocation */
+
+    UCT_AM_CB_FLAG_ASYNC = UCS_BIT(2), /**< callback may be invoked from any
```

Comment:
i think we should allow ASYNC callback to be used even if the interface does not support CB_ASYNC. For example, "put" callback is ASYNC, but it's (mostly) ok if it would be called only from uct_worker_progress